File: packages/client/src/lib/notification-content.ts
/**
 * 通知内容构建器。
 *
 * 架构位置：
 * - Cron/手动通知负责读取 settings 和订阅。
 * - 本模块只把领域数据转换成纯文本消息，不访问网络、不写数据库。
 *
 * 流程：
 * ```
 * now + settings + subscriptions -> date-only 比较 -> 文本分组 -> NotificationContent
 * ```
 */
import type { AppSettings, SubscriptionStatus } from "@/types/subscription";
import { daysBetweenDateOnly, isValidDateOnly, todayDateOnlyInTimeZone, type DateOnly } from "@/lib/time/date-only";
import { isValidTimeZone } from "@/lib/time/time-zone";
import { normalizeLocale, type Locale } from "@/i18n/locales";
import { translate } from "@/i18n/messages";

/**
 * 生成通知内容（不负责发送）。
 *
 * 说明：
 * - 该文件只做“输入 → 文本消息”的纯逻辑，便于后续写单测/复用到不同触发器（手动/定时）
 * - 金额不做汇率换算：通知更接近“原始扣费信息”（统计口径在页面里处理）
 *
 * Caveat: 这里使用 date-only 天数比较，不使用 Date 本地时区差值，避免服务器时区影响提醒日期。
 */

export interface SubscriptionForNotification {
  id: string;
  name: string;
  price: number;
  currency: string;
  status: SubscriptionStatus;
  nextBillingDate: string; // YYYY-MM-DD
  trialEndDate?: string | null; // YYYY-MM-DD | null
  /** 配置的提前提醒档位数组（每个值独立匹配 daysUntil）。 */
  reminderOffsets: number[];
}

export type NotificationItemType = "renewal" | "trial" | "expired";

/** 单个会进入通知内容的结构化条目，用于发送历史快照和即将提醒预览。 */
export interface NotificationContentItem {
  type: NotificationItemType;
  subscriptionId: string;
  name: string;
  price: number;
  currency: string;
  status: SubscriptionStatus;
  targetDate: string;
  /** 命中本条提醒的具体档位值（不是订阅配置的所有档位，仅触发本条的那一个）。 */
  reminderDays: number;
  daysUntil: number;
}

/** 通知内容输出；发送层只关心 title/content/timestamp，调度层使用 hasPayload 决定是否发送。 */
export interface NotificationContent {
  title: string;
  content: string;
  /** 用户可见的生成时间，已按用户选择的 IANA 时区格式化。 */
  timestamp: string;
  items: NotificationContentItem[];
  /** 用于判断“是否真的需要发送”的标志（例如没到期时不发）。 */
  hasPayload: boolean;
}

/** 取某个时区的“今天”（YYYY-MM-DD）。 */
export function getTodayDateOnlyInTimeZone(now: Date, timeZone: string): string {
  try {
    return todayDateOnlyInTimeZone(now, timeZone);
  } catch {
    return todayDateOnlyInTimeZone(now, "UTC");
  }
}

function resolveDisplayTimeZone(timeZone: string): string {
  const trimmed = timeZone.trim();
  return trimmed && isValidTimeZone(trimmed) ? trimmed : "UTC";
}

function getDateTimePart(parts: Intl.DateTimeFormatPart[], type: Intl.DateTimeFormatPartTypes): string {
  return parts.find((part) => part.type === type)?.value ?? "00";
}

/** 按用户选择的 IANA 时区格式化通知中展示给人的时间。 */
export function formatNotificationDisplayTime(now: Date, timeZone: string, locale: Locale = "zh-CN"): string {
  const displayTimeZone = resolveDisplayTimeZone(timeZone);
  const parts = new Intl.DateTimeFormat(locale, {
    timeZone: displayTimeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
    hourCycle: "h23",
  }).formatToParts(now);

  const year = getDateTimePart(parts, "year");
  const month = getDateTimePart(parts, "month");
  const day = getDateTimePart(parts, "day");
  const hour = getDateTimePart(parts, "hour");
  const minute = getDateTimePart(parts, "minute");
  const second = getDateTimePart(parts, "second");

  return `${year}-${month}-${day} ${hour}:${minute}:${second} ${displayTimeZone}`;
}

function formatAmount(amount: number): string {
  if (!Number.isFinite(amount)) return String(amount);
  const fixed = amount.toFixed(2);
  return fixed.replace(/\.00$/, "").replace(/(\.\d)0$/, "$1");
}

function formatItemLine(item: NotificationContentItem, locale: Locale): string {
  let extra: string;
  if (item.type === "trial") {
    extra = translate(locale, "notification.content.trialBeforeDays", { days: item.reminderDays });
  } else if (item.type === "expired") {
    extra = translate(locale, "notification.content.expiredStatus");
  } else {
    extra = translate(locale, "notification.content.beforeDays", { days: item.reminderDays });
  }
  if (locale === "en-US") {
    return `- ${item.name}: ${item.targetDate}, ${formatAmount(item.price)} ${item.currency} (${extra})`;
  }
  return `- ${item.name}：${item.targetDate}，${formatAmount(item.price)} ${item.currency}（${extra}）`;
}

function buildNotificationContentFromItems(
  now: Date,
  timeZone: string,
  items: NotificationContentItem[],
  locale: Locale,
): NotificationContent {
  const renewals = items.filter((item) => item.type === "renewal").map((item) => formatItemLine(item, locale));
  const trials = items.filter((item) => item.type === "trial").map((item) => formatItemLine(item, locale));
  const expired = items.filter((item) => item.type === "expired").map((item) => formatItemLine(item, locale));

  const blocks: string[] = [];
  if (renewals.length > 0) blocks.push([translate(locale, "notification.content.renewalBlock"), ...renewals].join("\n"));
  if (trials.length > 0) blocks.push([translate(locale, "notification.content.trialBlock"), ...trials].join("\n"));
  if (expired.length > 0) blocks.push([translate(locale, "notification.content.expiredBlock"), ...expired].join("\n"));

  const hasPayload = blocks.length > 0;
  const content = hasPayload
    ? blocks.join("\n\n")
    : translate(locale, "notification.content.empty");

  return {
    title: translate(locale, "notification.content.title"),
    content,
    timestamp: formatNotificationDisplayTime(now, timeZone, locale),
    items,
    hasPayload,
  };
}

/** 构造固定测试通知，用于验证单个渠道配置。 */
export function buildTestNotification(now: Date, timeZone: string, locale: Locale = "zh-CN"): NotificationContent {
  return {
    title: translate(locale, "notification.content.testTitle"),
    content: translate(locale, "notification.content.testBody"),
    timestamp: formatNotificationDisplayTime(now, timeZone, locale),
    items: [],
    hasPayload: true,
  };
}

/**
 * 计算某个用户本地日期会被纳入通知的条目。
 *
 * `includeExpired=false` 主要用于未来预览，避免把同一批已过期订阅重复塞进未来 30 天的每一天。
 */
export function collectNotificationItemsForLocalDate(
  localDate: DateOnly | string,
  settings: AppSettings,
  subscriptions: SubscriptionForNotification[],
  options: { includeExpired?: boolean } = {},
): NotificationContentItem[] {
  const includeExpired = options.includeExpired ?? true;
  const items: NotificationContentItem[] = [];

  for (const sub of subscriptions) {
    if (!isValidDateOnly(sub.nextBillingDate)) continue;
    const daysUntilNext = daysBetweenDateOnly(localDate, sub.nextBillingDate);
    // 过期条目展示用最大档位（仅用于排版/i18n，已过期不再参与档位匹配）。
    const displayOffsetForExpired = sub.reminderOffsets.length > 0
      ? Math.max(...sub.reminderOffsets)
      : 0;

    if (daysUntilNext < 0) {
      if (settings.showExpired && includeExpired) {
        items.push({
          type: "expired",
          subscriptionId: sub.id,
          name: sub.name,
          price: sub.price,
          currency: sub.currency,
          status: sub.status,
          targetDate: sub.nextBillingDate,
          reminderDays: displayOffsetForExpired,
          daysUntil: daysUntilNext,
        });
      }
    } else if (sub.reminderOffsets.includes(daysUntilNext)) {
      items.push({
        type: "renewal",
        subscriptionId: sub.id,
        name: sub.name,
        price: sub.price,
        currency: sub.currency,
        status: sub.status,
        targetDate: sub.nextBillingDate,
        reminderDays: daysUntilNext,
        daysUntil: daysUntilNext,
      });
    }

    if (sub.status === "trial" && sub.trialEndDate) {
      if (!isValidDateOnly(sub.trialEndDate)) continue;
      const daysUntilTrialEnd = daysBetweenDateOnly(localDate, sub.trialEndDate);
      if (daysUntilTrialEnd >= 0 && sub.reminderOffsets.includes(daysUntilTrialEnd)) {
        items.push({
          type: "trial",
          subscriptionId: sub.id,
          name: sub.name,
          price: sub.price,
          currency: sub.currency,
          status: sub.status,
          targetDate: sub.trialEndDate,
          reminderDays: daysUntilTrialEnd,
          daysUntil: daysUntilTrialEnd,
        });
      }
    }
  }

  return items;
}

/**
 * 生成”到期/试用结束”通知内容；没有需要提醒的订阅时返回 hasPayload=false。
 *
 * 规则：
 * - 续费提醒：`daysUntil(nextBillingDate) ∈ reminderOffsets`
 * - 试用结束提醒：`status=trial` 且 `daysUntil(trialEndDate) ∈ reminderOffsets`
 * - 已过期订阅（可选）：`nextBillingDate < today` 且 settings.showExpired=true
 * - 命中多个档位时按当日合并：一条消息列出全部命中的订阅 + 各自档位
 */
export function buildDueNotification(
  now: Date,
  settings: AppSettings,
  subscriptions: SubscriptionForNotification[],
): NotificationContent {
  const today = getTodayDateOnlyInTimeZone(now, settings.timezone || "UTC");
  return buildDueNotificationForLocalDate(today, now, settings, subscriptions);
}

/**
 * 按指定用户本地调度日构建通知内容。
 *
 * Cron 可能在跨午夜的容错窗口内补跑上一天的计划任务，因此不能总是用 `now`
 * 推导本地日期；否则 23:59 的计划在 00:01 补跑时会错过上一天应命中的订阅。
 */
export function buildDueNotificationForLocalDate(
  localDate: DateOnly | string,
  now: Date,
  settings: AppSettings,
  subscriptions: SubscriptionForNotification[],
): NotificationContent {
  const items = collectNotificationItemsForLocalDate(localDate, settings, subscriptions);
  return buildNotificationContentFromItems(now, settings.timezone || "UTC", items, normalizeLocale(settings.locale));
}

File: packages/server/cmd/renewlet/notification_content.go
package main

// notification_content.go 将订阅投影转换为可发送的通知内容。
//
// 架构位置：调度器、手动运行和测试发送共享同一套内容构建，确保历史记录、渠道文本和前端预览口径一致。
// 这里刻意按 date-only 计算提醒窗口，因为扣费日是用户本地业务日期，不应被 UTC instant 或 DST 影响。
//
// Caveat: 调整 item type 或文案分组会影响所有渠道文本和 notification job result schema。
import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

// listNotificationSubscriptions 读取通知计算所需的订阅投影。
func listNotificationSubscriptions(app core.App, userID string) ([]notificationSubscription, error) {
	rows, err := app.FindAllRecords("subscriptions", dbx.HashExp{"user": userID})
	if err != nil {
		return nil, err
	}
	subscriptions := make([]notificationSubscription, 0, len(rows))
	for _, row := range rows {
		offsets, err := normalizeReminderOffsetsValue(row.Get("reminderOffsets"), row.GetInt("reminderDays"))
		if err != nil {
			// 历史脏数据兜底：退化为单档位，避免单条记录把整批通知拉崩。
			offsets = []int{row.GetInt("reminderDays")}
		}
		subscriptions = append(subscriptions, notificationSubscription{
			ID:              row.Id,
			Name:            row.GetString("name"),
			LogoURL:         row.GetString("logo"),
			Price:           row.GetFloat("price"),
			Currency:        row.GetString("currency"),
			Status:          row.GetString("status"),
			NextBillingDate: row.GetString("nextBillingDate"),
			TrialEndDate:    row.GetString("trialEndDate"),
			ReminderOffsets: offsets,
		})
	}
	return subscriptions, nil
}

// buildTestNotification 构造测试通知内容。
func buildTestNotification(now time.Time, settings appSettings) notificationMessage {
	locale := normalizeAppLocale(settings.Locale)
	return notificationMessage{
		Title:      tr(locale, "Renewlet 测试通知", "Renewlet test notification"),
		Content:    tr(locale, "如果你收到了这条消息，说明该通知渠道配置可用。", "If you received this message, this notification channel is working."),
		Timestamp:  formatNotificationTime(now, settings.Timezone),
		Items:      []notificationContentItem{},
		HasPayload: true,
	}
}

// buildDueNotification 根据当前时间和用户时区构造到期提醒。
func buildDueNotification(now time.Time, settings appSettings, subscriptions []notificationSubscription, includeExpired bool) notificationMessage {
	localDate := todayDateOnly(now, settings.Timezone)
	items := collectNotificationItems(localDate, settings, subscriptions, includeExpired)
	return buildNotificationContent(now, settings, items)
}

// buildDueNotificationForLocalDate 按指定本地日期构造提醒。
func buildDueNotificationForLocalDate(localDate string, now time.Time, settings appSettings, subscriptions []notificationSubscription, includeExpired bool) notificationMessage {
	items := collectNotificationItems(localDate, settings, subscriptions, includeExpired)
	return buildNotificationContent(now, settings, items)
}

// collectNotificationItems 收集指定本地日期应该提醒的项目。
// 为什么用 date-only 差值：订阅扣费日是业务日期，不应受 UTC instant 或 DST 切换影响。
func collectNotificationItems(localDate string, settings appSettings, subscriptions []notificationSubscription, includeExpired bool) []notificationContentItem {
	items := []notificationContentItem{}
	for _, sub := range subscriptions {
		// 过期条目展示用最大档位（仅用于排版/i18n，过期不再参与档位匹配）。
		displayOffsetForExpired := 0
		if len(sub.ReminderOffsets) > 0 {
			displayOffsetForExpired = sub.ReminderOffsets[0] // 已降序排列
		}
		if isValidDateOnly(sub.NextBillingDate) {
			daysUntilNext := daysBetweenDateOnly(localDate, sub.NextBillingDate)
			if daysUntilNext < 0 {
				if settings.ShowExpired && includeExpired {
					items = append(items, notificationContentItem{
						Type:           "expired",
						SubscriptionID: sub.ID,
						Name:           sub.Name,
						LogoURL:        sub.LogoURL,
						Price:          sub.Price,
						Currency:       sub.Currency,
						Status:         normalizeSubscriptionStatus(sub.Status),
						TargetDate:     sub.NextBillingDate,
						ReminderDays:   displayOffsetForExpired,
						DaysUntil:      daysUntilNext,
					})
				}
			} else if reminderOffsetsContains(sub.ReminderOffsets, daysUntilNext) {
				items = append(items, notificationContentItem{
					Type:           "renewal",
					SubscriptionID: sub.ID,
					Name:           sub.Name,
					LogoURL:        sub.LogoURL,
					Price:          sub.Price,
					Currency:       sub.Currency,
					Status:         normalizeSubscriptionStatus(sub.Status),
					TargetDate:     sub.NextBillingDate,
					ReminderDays:   daysUntilNext,
					DaysUntil:      daysUntilNext,
				})
			}
		}

		if sub.Status == "trial" && isValidDateOnly(sub.TrialEndDate) {
			daysUntilTrialEnd := daysBetweenDateOnly(localDate, sub.TrialEndDate)
			if daysUntilTrialEnd >= 0 && reminderOffsetsContains(sub.ReminderOffsets, daysUntilTrialEnd) {
				items = append(items, notificationContentItem{
					Type:           "trial",
					SubscriptionID: sub.ID,
					Name:           sub.Name,
					LogoURL:        sub.LogoURL,
					Price:          sub.Price,
					Currency:       sub.Currency,
					Status:         "trial",
					TargetDate:     sub.TrialEndDate,
					ReminderDays:   daysUntilTrialEnd,
					DaysUntil:      daysUntilTrialEnd,
				})
			}
		}
	}
	return items
}

// reminderOffsetsContains 判断订阅配置的提醒档位是否包含某个 daysUntil。
func reminderOffsetsContains(offsets []int, value int) bool {
	for _, offset := range offsets {
		if offset == value {
			return true
		}
	}
	return false
}

// buildNotificationContent 将提醒项分组为可读消息。
func buildNotificationContent(now time.Time, settings appSettings, items []notificationContentItem) notificationMessage {
	locale := normalizeAppLocale(settings.Locale)
	renewals := []string{}
	trials := []string{}
	expired := []string{}
	for _, item := range items {
		line := formatNotificationItemLine(item, locale)
		switch item.Type {
		case "trial":
			trials = append(trials, line)
		case "expired":
			expired = append(expired, line)
		default:
			renewals = append(renewals, line)
		}
	}

	blocks := []string{}
	if len(renewals) > 0 {
		blocks = append(blocks, tr(locale, "即将续费：", "Upcoming renewals:")+"\n"+strings.Join(renewals, "\n"))
	}
	if len(trials) > 0 {
		blocks = append(blocks, tr(locale, "试用结束：", "Trial ending:")+"\n"+strings.Join(trials, "\n"))
	}
	if len(expired) > 0 {
		blocks = append(blocks, tr(locale, "已过期（未更新下次扣费日期）：", "Expired (next billing date not updated):")+"\n"+strings.Join(expired, "\n"))
	}
	hasPayload := len(blocks) > 0
	content := tr(locale, "今天没有需要提醒的订阅（你可以在设置页关闭“每日通知”，或调整各订阅的提醒天数）。", "No subscriptions need reminders today. You can disable daily notifications in Settings or adjust reminder days for subscriptions.")
	if hasPayload {
		content = strings.Join(blocks, "\n\n")
	}
	return notificationMessage{
		Title:      tr(locale, "Renewlet 订阅提醒", "Renewlet subscription reminder"),
		Content:    content,
		Timestamp:  formatNotificationTime(now, settings.Timezone),
		Items:      items,
		HasPayload: hasPayload,
	}
}

func formatNotificationItemLine(item notificationContentItem, locale appLocale) string {
	extra := fmt.Sprintf(tr(locale, "提前 %d 天提醒", "%d days before"), item.ReminderDays)
	if item.Type == "trial" {
		extra = fmt.Sprintf(tr(locale, "试用结束，提前 %d 天提醒", "trial ends, %d days before"), item.ReminderDays)
	} else if item.Type == "expired" {
		extra = tr(locale, "已过期", "expired")
	}
	if locale == localeEnUS {
		return fmt.Sprintf("- %s: %s, %s %s (%s)", item.Name, item.TargetDate, formatAmount(item.Price), item.Currency, extra)
	}
	return fmt.Sprintf("- %s：%s，%s %s（%s）", item.Name, item.TargetDate, formatAmount(item.Price), item.Currency, extra)
}

func formatAmount(amount float64) string {
	if math.IsNaN(amount) || math.IsInf(amount, 0) {
		return fmt.Sprintf("%v", amount)
	}
	fixed := strconv.FormatFloat(amount, 'f', 2, 64)
	fixed = strings.TrimSuffix(fixed, ".00")
	if strings.HasSuffix(fixed, "0") && strings.Contains(fixed, ".") {
		fixed = strings.TrimSuffix(fixed, "0")
	}
	return fixed
}

func formatNotificationTime(now time.Time, timezone string) string {
	loc, err := time.LoadLocation(timezone)
	if err != nil {
		loc = time.UTC
		timezone = "UTC"
	}
	return now.In(loc).Format("2006-01-02 15:04:05") + " " + timezone
}

func normalizeSubscriptionStatus(status string) string {
	switch status {
	case "trial", "active", "paused", "cancelled":
		return status
	default:
		return "active"
	}
}
